### PR TITLE
Automated source code style checking

### PR DESCRIPTION
**Pull request by smharper**
This PR makes three broad changes: it
* adds a script, /tests/check_source.py, that parses Fortran source code and looks for style errors,
* adds that script to the Travis CI testing,
* and changes the source code to comply with said script

The script checks for
* the presence of tab characters,
* trailing whitespace,
* lines that contain only whitespace,
* indentation (for normal lines, it simply checks to see if there are an even number of indentation spaces; for continuation lines, it makes sure they are indented at least five spaces beyond the previous indentation level)

To see what the output looks like for code that doesn't match our style, check out this CI build: https://travis-ci.org/smharper/openmc/jobs/72062690

### CODE REVIEW

File: tests/check_source.py [hunk 1]
```diff
@@ -0,0 +1,235 @@
...
+class LineOfCode(object):
+    """Contains and provides basic info about a line of Fortran 90 code."""
+    def __init__(self, number, content, is_continuation=False,
+                 string_terminator=None):
+        # Initialize member variables.
+        self.number = number
+        self.content = content
+        self.is_continuation = is_continuation
+        self.is_continued = False
+        assert (string_terminator == None or
```
Comment (paulromano) [inline tests/check_source.py]: Should be `string_terminator is None`

File: tests/check_source.py [hunk 2]
```diff
@@ -0,0 +1,235 @@
...
+            self.final_string_terminator = "'"
+        if in_a_double_quote:
+            assert self.is_continued
+            self.final_string_terminator = '"'
+
+        # Is this line a comment-only line?
+        if (not has_real_content) and in_a_comment:
+            self.is_comment_only = True
+
+
```
Comment (paulromano) [inline tests/check_source.py]: PEP8: methods separated by one blank line

File: tests/check_source.py [hunk 3]
```diff
@@ -0,0 +1,235 @@
...
+    def contains_whitespace_only(self):
+        """Return True/False if all characters in the line are whitespace."""
+        if len(self.content.strip('\n')) == 0: return False
+        is_ws = [ any([char == ws for ws in whitespace])
+                  for char in self.content.strip('\n') ]
+        return all(is_ws)
+
+
```
Comment (paulromano) [inline tests/check_source.py]: PEP8: methods separated by one blank line

File: tests/check_source.py [hunk 4]
```diff
@@ -0,0 +1,235 @@
...
+        # Is this line a comment-only line?
+        if (not has_real_content) and in_a_comment:
+            self.is_comment_only = True
+
+
+    def get_indent(self):
+        """Return the number of indentation spaces prefixing this line."""
+        return len(self.content) - len(self.content.lstrip(' '))
+
+
```
Comment (paulromano) [inline tests/check_source.py]: PEP8: methods separated by one blank line

File: tests/check_source.py [hunk 5]
```diff
@@ -0,0 +1,235 @@
...
+    otherwise.
+    """
+    good_code = True
+    base_indent = 0
+
+    for loc in read_lines_of_code(fname):
+        # Check for extra whitespace errors.
+        if loc.contains_whitespace_only():
+            good_code = False
+            print_error(fname, loc.number, "Line contains whitespace " \
```
Comment (paulromano) [inline tests/check_source.py]: Backslash is redundant here

File: tests/check_source.py [hunk 6]
```diff
@@ -0,0 +1,235 @@
...
+                 "characters but no content.  Please remove whitespace.")
+        elif loc.has_trailing_whitespace():
+            good_code = False
+            print_error(fname, loc.number, "Line has trailing whitespace after"\
+                 " the content.  Please remove trailing whitespace.")
+        if loc.contains_tab():
+            good_code = False
+            print_error(fname, loc.number, "Line contains a tab character.  " \
```
Comment (paulromano) [inline tests/check_source.py]: Backslash is redundant here

File: tests/check_source.py [hunk 7]
```diff
@@ -0,0 +1,235 @@
...
+                 "Please replace with single whitespace characters.")
+
+        # Check indentation.
+        current_indent = loc.get_indent()
+        if ((not loc.is_continuation) and (not loc.is_comment_only) and
+              (not loc.contains_whitespace_only()) and current_indent % 2 != 0):
+            good_code = False
+            print_error(fname, loc.number, "Line is indented an odd number of "\
+                 "spaces.  All non-continuation lines should be indented an "\
```
Comment (paulromano) [inline tests/check_source.py]: Backslash is redundant here

File: tests/check_source.py [hunk 8]
```diff
@@ -0,0 +1,235 @@
...
+                 "even number of spaces.")
+        if loc.is_continuation and current_indent < base_indent + 5:
+            good_code = False
+            print_error(fname, loc.number, "Continuation lines must be "\
+                 "indented by at least 5 spaces, but this line is indented {0}"\
```
Comment (paulromano) [inline tests/check_source.py]: Backslash is redundant here

**Comment by paulromano** [comment]
Very cool script @smharper! Can we make the trailing whitespace and line with only whitespace messages warnings instead of errors? Our style guide doesn't actually say thou shalt never have trailing whitespace or empty lines with whitespace. Nevertheless, it's good to have that check during a PR.

**Comment by smharper** [comment]
I've addressed the `is None`, extra newline, and redundant backslash issues.

About the trailing whitespace and whitespace only errors: Are you open to the idea of forbidding them in the style guide?

**Comment by wbinventor** [comment]
I like the idea of forbidding them in the style guide now that this script exists.

**Comment by paulromano** [comment]
Sorry for the delayed response -- yes, I'm open to forbidding them in the style guide as well. Whitespace only is really just a special case of trailing whitespace, so forbidding trailing whitespace alone should suffice.

**Comment by smharper** [comment]
The style guide is now appropriately modified.

**Comment by paulromano** [comment]
@bhermanmit @nelsonag @walshjon @nhorelik Any qualms with the proposed changes?

**Comment by walshjon** [comment]
I don't have any problems with the proposed changes. Everything looks good to me. Nice idea with this PR, @smharper!

**Comment by nelsonag** [comment]
No further comments, this is great work!

**Pull request merged**